test/tstUnionFind.cpp

```diff
@@ -0,0 +1,124 @@
+/****************************************************************************
+ * Copyright (c) 2017-2022 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+#include "ArborX_EnableDeviceTypes.hpp" // ARBORX_DEVICE_TYPES
+#include "ArborX_EnableViewComparison.hpp"
+#include <ArborX_DetailsUnionFind.hpp>
+#include <ArborX_DetailsUtils.hpp> // iota
+
+#include "BoostTest_CUDA_clang_workarounds.hpp"
+#include <boost/test/unit_test.hpp>
+
+BOOST_AUTO_TEST_SUITE(UnionFind)
+
+namespace
+{
+
+template <typename ExecutionSpace, typename UnionFind>
+Kokkos::View<int *, Kokkos::HostSpace>
+build_representatives(ExecutionSpace const &space, UnionFind union_find)
+{
+  using MemorySpace = typename UnionFind::memory_space;
+
+  auto const n = union_find.size();
+  Kokkos::View<int *, MemorySpace> representatives("Test::representatives", n);
+  Kokkos::View<int *, MemorySpace> map2smallest("Test::map", n);
+  Kokkos::deep_copy(space, map2smallest, INT_MAX);
+  Kokkos::parallel_for(
+      "Test::find_representatives",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0, n), KOKKOS_LAMBDA(int i) {
+        auto r = union_find.representative(i);
+#if KOKKOS_VERSION >= 30799
+        Kokkos::atomic_min(&map2smallest(r), i);
+#else
+         // Workaround for undefined
+        //   desul::atomic_min(int*, int, desul::MemoryOrderRelaxed,
+        //   desul::MemoryScopeDevice)
+        // in older Kokkos versions.
+        auto v = map2smallest(r);
+        while (v > i) {
+          v = Kokkos::atomic_compare_exchange(&map2smallest(r), v, i);
+        }
+#endif
+        representatives(i) = r;
+      });
+  // We want the representative values to not depend on a specific
+  // implementation of the union-find (e.g., not relying them being equal to
+  // the smallest index in the set), so we explicitly remap them.
+  Kokkos::parallel_for(
+      "Test::remap_representatives",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0, n), KOKKOS_LAMBDA(int i) {
+        representatives(i) = map2smallest(representatives(i));
+      });
```

Comment:
I struggle to convince myself that this does the right thing w/o making a temporary copy of `representative`

Reply:
This reads and then writes one value in a view. Why would you want a copy?

Reply:
Nevermind